source/utils/adios_remote_read/adios_remote_read.cpp

```diff
@@ -0,0 +1,721 @@
+            {
+                out += "\n";
+                counter = 0;
+            }
+        }
+    }
+    return out;
+}
+
+void read3D(int nproc, int rank, const std::string &filename, const int nsteps, adios2::IO &io,
+            std::vector<std::string> &required_variables, int direction, double ratio,
+            int output_line_length)
+{
+    unsigned int startX;
+    unsigned int startY;
+    unsigned int startZ;
+    unsigned int countX;
+    unsigned int countY;
+    unsigned int countZ;
+
+    try
+    {
+        std::chrono::time_point<std::chrono::system_clock> start_time;
+        std::chrono::time_point<std::chrono::system_clock> end_time;
+        start_time = std::chrono::system_clock::now();
+
+        adios2::Engine reader = io.Open(filename, adios2::Mode::Read);
+
+        std::string out;
+        for (size_t step = 0; step < nsteps; step++)
+        {
+            reader.BeginStep();
+            auto available_variables = io.AvailableVariables(true);
+            std::vector<std::string> available_variables_str;
+            for (const auto &v : available_variables)
+                available_variables_str.push_back(v.first);
+            std::vector<std::string> vars_intersection;
+            if (!required_variables.empty())
+            {
+                std::sort(available_variables_str.begin(), available_variables_str.end());
+
+                std::set_intersection(available_variables_str.begin(),
+                                      available_variables_str.end(), required_variables.begin(),
+                                      required_variables.end(),
+                                      std::back_inserter(vars_intersection));
+            }
+            else
+            {
+                vars_intersection = available_variables_str;
+            }
+            for (auto const &var_name : vars_intersection)
+            {
+                adios2::Variable<double> var = io.InquireVariable<double>(var_name);
+                if (var && var.Shape().size() == 3)
+                {
+                    auto globalSizeX = var.Shape()[0];
+                    auto globalSizeY = var.Shape()[1];
+                    auto globalSizeZ = var.Shape()[2];
+
+                    if (ratio == 1.0)
+                    {
+                        startX = 0;
+                        countX = globalSizeX;
+
+                        startY = 0;
+                        countY = globalSizeY;
+
+                        startZ = 0;
+                        countZ = globalSizeZ;
+                    }
+                    else
+                    {
+                        startX = (1 - ratio) / 2.0 * globalSizeX;
+                        countX = globalSizeX * ratio;
+
+                        startY = (1 - ratio) / 2.0 * globalSizeY;
+                        countY = globalSizeY * ratio;
+
+                        startZ = (1 - ratio) / 2.0 * globalSizeZ;
+                        countZ = globalSizeZ * ratio;
```

Comment:
What exactly is this `ratio` ? (maybe you should add a comment to explain below where you define the variables)
It seems you want to control the amount you are reading (ratio of the total data on that axes). My concern is with the start. For ratio of 0.5 you are bringing half the data starting from position `globalSize`/4 
Is this just a random start point or it represents something?

Reply:
Yes, the ration parameter corresponds to the amount of data that should be read. Probably it is the most short description of the task. Specifying starts and counts for every MPI rank is not really convenient for the user.

Reply:
Yes, now that I looked at the paper it makes sense. My confusion was with the start point, but it's just an arbitrary choice.